src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -230,13 +230,14 @@ internal override bool GetMemberAndWriteJson(object obj, ref WriteStack state, U
 
             if (IgnoreDefaultValuesOnWrite)
             {
-                // If value is null, it is a reference type or nullable<T>.
-                if (value == null)
+                if (default(T) is null)
                 {
-                    return true;
+                    if (value is null)
+                    {
+                        return true;
+                    }
                 }
-
-                if (!PropertyTypeCanBeNull)
+                else
```

Comment:
wasn't previous logic correct? When `value` is null then it can be null so there is no need to check for `default(T) is null`? Now we're doing extra check

Reply:
It's the other way around. Also, as discussed `default(T) is null` is a constant as far as the JIT is concerned so it will eliminate the branching.

Reply:
fair enough